cli/dcos-spark/submit_builder.go

```diff
@@ -336,12 +347,36 @@ ARGLOOP:
 					continue ARGLOOP
 				}
 			}
-			// merge this --key against the following val to get --key=val
-			argsEquals = append(argsEquals, arg+"="+args[i+1])
+
+			// if this is the beginning of a string of args e.g. '-Djava.option=setting -Djava.paramter=nonsense'
+			// we want to remove the leading single quote. Also remove internal quotes when the arg == --conf or some
+			// other named configuration
+			// e.g.: next = spark.driver.extraJavaOptions='-Djava.something=somethingelse
+			// arg =  --conf
+			arg = strings.TrimPrefix(arg, "'")
+			next := args[i + 1]
+			if strings.HasPrefix(next, "'") {  // e.g. --driver-java-options '-Djava.config=setting... <-- next
+				inQuotes = true
+			}
+			next = strings.Replace(next, "'", "", -1)  // remove internal quotes
+			argsEquals = append(argsEquals, arg + "=" + next)
 			i += 2
+		} else if strings.HasSuffix(arg, "'") {  // attach the final arg to the string of args without the quote
+			inQuotes = false  // has suffix means we're out of the quotes
+			arg = strings.TrimSuffix(arg, "'")
+			argsEquals[len(argsEquals) - 1] = argsEquals[len(argsEquals) - 1] + " " + arg
+			i += 1
 		} else {
-			// already joined or at the end, pass through:
-			argsEquals = append(argsEquals, arg)
+			cleanedArg := strings.Replace(arg, "'", "", -1)
+			if inQuotes { // join this arg to the last one because it's all in quotes
+				argsEquals[len(argsEquals) - 1] = argsEquals[len(argsEquals) - 1] + " " + cleanedArg
+			} else {
+				if strings.Contains(arg, "'") {  // e.g. --driver-java-options='-Djava.firstConfig=firstSetting
+					inQuotes = true
+				}
+				// already joined or at the end, pass through
+				argsEquals = append(argsEquals, cleanedArg)
+			}
```

Comment:
Maybe we can document somewhere that we expect single quotes to be used when a value has spaces in it.